File: examples/fully_sampled_recon_single_chain.m
% Lower-level interface demo, creates and runs a chain of gadgets.

if ~libisloaded('mutilities')
    fprintf('loading mutilities library...\n')
    [notfound, warnings] = loadlibrary('mutilities');
end
if ~libisloaded('mgadgetron')
    fprintf('loading mgadgetron library...\n')
    [notfound, warnings] = loadlibrary('mgadgetron');
end

%libfunctions('mutilities')
%libfunctions('mgadgetron')

try
    % define gadgets
    gadget1 = gadgetron.Gadget('RemoveROOversamplingGadget');
	gadget2 = gadgetron.Gadget('AcquisitionAccumulateTriggerGadget');
	gadget3 = gadgetron.Gadget('BucketToBufferGadget');
	gadget4 = gadgetron.Gadget('SimpleReconGadget');
	gadget5 = gadgetron.Gadget('ImageArraySplitGadget');
	gadget6 = gadgetron.Gadget('ExtractGadget');
    
    % set gadget parameters
    gadget2.set_property('trigger_dimension', 'repetition')
    gadget3.set_property('split_slices', 'true')
    
    % create reconstructor
    recon = gadgetron.ImagesReconstructor();

    % build gadget chain
    recon.add_gadget('g1', gadget1);
	recon.add_gadget('g2', gadget2);
	recon.add_gadget('g3', gadget3);
	recon.add_gadget('g4', gadget4);
	recon.add_gadget('g5', gadget5);
	recon.add_gadget('g6', gadget6);
    
    % define raw data source
    input_data = gadgetron.MR_Acquisitions('testdata.h5');    
    recon.set_input(input_data)
    % perform reconstruction
    recon.process()
    % get reconstructed images
    images = recon.get_output();
    
    % plot reconstructed images
    for i = 1 : images.number()
        data = images.image_as_array(i);
        figure(i)
        data = data/max(max(max(data)));
        imshow(data(:,:,1));
    end

    % write images to a new group in 'output1.h5'
    % named after the current date and time
    fprintf('appending output1.h5...\n')
    images.write('output1.h5', datestr(datetime))

catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end

File: xGadgetron/examples/fully_sampled_recon_single_chain.py
'''
Medium-level interface demo that illustrates 2D Cartesian MR image 
reconstruction using Gadgetron by directly creating and running a chain of 
gadgets.

Usage:
  fully_sampled_recon_single_chain.py [--help | options]

Options:
  -f <file>, --file=<file>    raw data file
                              [default: simulated_MR_2D_cartesian.h5]
  -p <path>, --path=<path>    sub-path to engine module
                              [default: /xGadgetron/pGadgetron]
  -o <file>, --output=<file>  images output file
'''

__version__ = '0.1.0'
from docopt import docopt
args = docopt(__doc__, version=__version__)

import os
import sys
import time

sys.path.append(os.environ.get('SRC_PATH') + args['--path'])

from pGadgetron import *

def main():

    # acquisitions will be read from an HDF file
    input_data = AcquisitionData(args['--file'])
    
    # create reconstruction object
    # Rather than using a predefined image reconstruction object, here a new 
    # image reconstruction object is created by concatinating multiple gadgets 
    # (for more information on Gadgetron and its gadgets please see: 
    # https://github.com/gadgetron/.).
    # Parameters for individual gadgets can be defined either during the 
    # creation of the reconstruction object:
    #   e.g. AcquisitionAccumulateTriggerGadget(trigger_dimension=repetition)
    # or by giving a gadget a label (cf. label ex: for the last gadget)
    # and using set_gadget_property(label, propery, value).
    # The gadgets will be concatenated and will be executed as soon as 
    # process() is called.
    recon = ImagesReconstructor(['RemoveROOversamplingGadget', \
        'AcquisitionAccumulateTriggerGadget(trigger_dimension=repetition)', \
        'BucketToBufferGadget(split_slices=true, verbose=false)', \
        'SimpleReconGadget', 'ImageArraySplitGadget', 'ex:ExtractGadget'])
        
    # ExtractGadget defines which type of image should be returned:
    # none      0
    # magnitude 1
    # real      2
    # imag      4
    # phase     8
    # max       16  
    # in this example '5' returns both magnitude and imag    
    recon.set_gadget_property('ex', 'extract_mask', 5) 
    
    # provide raw k-space data as input
    recon.set_input(input_data)
    
    # perform reconstruction
    recon.process()
    
    # retrieve reconstructed images
    images = recon.get_output()

    # show reconstructed images
    images.show()

    if args['--output'] is not None:
        # write images to a new group in args.output
        # named after the current date and time
        time_str = time.asctime()
        print('writing to %s' % args['--output'])
        images.write(args['--output'], time_str)

try:
    main()
    print('done')

except error as err:
    # display error information
    print('??? %s' % err.value)
